File: expandabletextview-lib/src/main/java/ru/dimakron/expandabletextview_lib/ExpandableTextView.java
package ru.dimakron.expandabletextview_lib;


import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Color;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.TextPaint;
import android.text.method.LinkMovementMethod;
import android.text.style.ClickableSpan;
import android.util.AttributeSet;
import android.view.View;

import androidx.appcompat.widget.AppCompatTextView;
import androidx.core.content.ContextCompat;

import org.jetbrains.annotations.NotNull;

public class ExpandableTextView extends AppCompatTextView {

    private static final int DEFAULT_TRIM_LENGTH = 240;
    private static final boolean DEFAULT_SHOW_TRIM_EXPANDED_TEXT = true;
    private static final String ELLIPSIZE = "... ";

    private CharSequence text;
    private BufferType bufferType;
    private boolean readMore = true;
    private int trimLength;
    private CharSequence trimCollapsedText;
    private CharSequence trimExpandedText;
    private ReadMoreClickableSpan viewMoreSpan;
    private int colorClickableText;
    private boolean showTrimExpandedText;

    private TrimTextInterceptor trimTextInterceptor;

    public ExpandableTextView(Context context) {
        this(context, null);
    }

    public ExpandableTextView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ExpandableTextView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);

        TypedArray typedArray = context.obtainStyledAttributes(attrs, R.styleable.ExpandableTextView);
        this.trimLength = typedArray.getInt(R.styleable.ExpandableTextView_trimLength, DEFAULT_TRIM_LENGTH);

        int resourceIdTrimCollapsedText = typedArray.getResourceId(R.styleable.ExpandableTextView_trimCollapsedText, R.string.read_more_show);
        int resourceIdTrimExpandedText = typedArray.getResourceId(R.styleable.ExpandableTextView_trimExpandedText, R.string.read_more_hide);

        this.trimCollapsedText = getResources().getString(resourceIdTrimCollapsedText);
        this.trimExpandedText = getResources().getString(resourceIdTrimExpandedText);
        this.colorClickableText = typedArray.getColor(R.styleable.ExpandableTextView_colorClickableText, ContextCompat.getColor(context, R.color.link));
        this.showTrimExpandedText = typedArray.getBoolean(R.styleable.ExpandableTextView_showTrimExpandedText, DEFAULT_SHOW_TRIM_EXPANDED_TEXT);
        typedArray.recycle();
        viewMoreSpan = new ReadMoreClickableSpan();
        setText();
    }

    private void setText() {
        super.setText(getDisplayableText(), bufferType);
        setMovementMethod(LinkMovementMethod.getInstance());
        setHighlightColor(Color.TRANSPARENT);
    }

    private CharSequence getDisplayableText() {
        return getTrimmedText(text);
    }

    @Override
    public void setText(CharSequence text, BufferType type) {
        this.text = text;
        bufferType = type;
        setText();
    }

    private CharSequence getTrimmedText(CharSequence text) {
        if (text != null && text.length() > trimLength) {
            if (readMore) {
                return updateCollapsedText();
            } else {
                return updateExpandedText();
            }
        }
        return text;
    }

    private CharSequence updateCollapsedText() {
        int trimEndIndex = trimLength + 1;
        SpannableStringBuilder s = new SpannableStringBuilder(text, 0, trimEndIndex)
                .append(ELLIPSIZE)
                .append(trimCollapsedText);
        return addClickableSpan(s, trimCollapsedText);
    }

    private CharSequence updateExpandedText() {
        if (showTrimExpandedText) {
            SpannableStringBuilder s = new SpannableStringBuilder(text, 0, text.length()).append(trimExpandedText);
            return addClickableSpan(s, trimExpandedText);
        }
        return text;
    }

    private CharSequence addClickableSpan(SpannableStringBuilder s, CharSequence trimText) {
        s.setSpan(viewMoreSpan, s.length() - trimText.length(), s.length(), Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);

        if(trimTextInterceptor != null) trimTextInterceptor.interceptTrimText(s, trimText);

        return s;
    }

    public void setTrimLength(int trimLength) {
        this.trimLength = trimLength;
        setText();
    }

    public void setColorClickableText(int colorClickableText) {
        this.colorClickableText = colorClickableText;
    }

    public void setTrimCollapsedText(CharSequence trimCollapsedText) {
        this.trimCollapsedText = trimCollapsedText;
    }

    public void setTrimExpandedText(CharSequence trimExpandedText) {
        this.trimExpandedText = trimExpandedText;
    }

    public void setTrimTextInterceptor(TrimTextInterceptor trimTextInterceptor){
        this.trimTextInterceptor = trimTextInterceptor;
        setText();
    }

    public interface TrimTextInterceptor{
        void interceptTrimText(SpannableStringBuilder s, CharSequence trimText);
    }

    private class ReadMoreClickableSpan extends ClickableSpan {
        @Override
        public void onClick(@NotNull View widget) {
            readMore = !readMore;
            setText();
        }

        @Override
        public void updateDrawState(@NotNull TextPaint ds) {
            ds.setColor(colorClickableText);
        }
    }
}

File: expandabletextview-lib/src/main/java/ru/dimakron/expandabletextview_lib/ExpandableTextView.kt
package ru.dimakron.expandabletextview_lib

import android.content.Context
import android.graphics.Color
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.TextPaint
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.util.AttributeSet
import android.view.View
import androidx.appcompat.widget.AppCompatTextView

class ExpandableTextView: AppCompatTextView, View.OnClickListener {

    private var rawText: CharSequence = ""
    private var collapsedText: CharSequence = ""
    private var expandedText: CharSequence = ""

    private var bufferType = BufferType.NORMAL
    private var isExpanded = false

    var trimLength = 240
        set(value) {
            field = value
            update()
        }

    var collapsedLinkText: CharSequence = ""
        set(value) {
            field = value
            update()
        }

    var expandedLinkText: CharSequence = ""
        set(value) {
            field = value
            update()
        }

    var linkColor = Color.BLUE
        set(value) {
            field = value
            update()
        }

    var toggleByLinkOnly = true
        set(value) {
            field = value
            update()
        }

    var onStateChange: ((Boolean) -> Unit)? = null

    var onCustomizeSpannable: ((s: Spannable, linkText: CharSequence) -> Unit)? = null
        set(value) {
            field = value
            update()
        }

    constructor(context: Context): super(context) {
        init()
    }

    constructor(context: Context, attrs: AttributeSet?): super(context, attrs) {
        init(attrs)
    }

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int): super(context, attrs, defStyleAttr) {
        init(attrs)
    }

    private fun init(attrs: AttributeSet? = null) {
        val a = attrs?.let { context.theme.obtainStyledAttributes(it, R.styleable.ExpandableTextView, 0, 0) }
        try {
            trimLength = a?.getInt(R.styleable.ExpandableTextView_etv_trimLength, trimLength)?: trimLength
            collapsedLinkText = a?.getString(R.styleable.ExpandableTextView_etv_textCollapsedLink)?: resources.getString(R.string.read_more_show)
            expandedLinkText = a?.getString(R.styleable.ExpandableTextView_etv_textExpandedLink)?: resources.getString(R.string.read_more_hide)
            linkColor = a?.getColor(R.styleable.ExpandableTextView_etv_textColorLink, linkColor)?: linkColor
            toggleByLinkOnly = a?.getBoolean(R.styleable.ExpandableTextView_etv_toggleByLinkOnly, toggleByLinkOnly)?: toggleByLinkOnly
        } finally {
            a?.recycle()
        }

        update()
    }

    private fun update() {
        if(trimLength >= rawText.length){
            collapsedText = rawText
            expandedText = rawText
        } else {
            val collapsedSpannable = SpannableStringBuilder()
                    .append(rawText.subSequence(0, trimLength))
                    .append("... ")
                    .append(collapsedLinkText)
            if(toggleByLinkOnly) {
                collapsedSpannable.setSpan(clickableSpan, collapsedSpannable.length - (collapsedLinkText.length), collapsedSpannable.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
            onCustomizeSpannable?.invoke(collapsedSpannable, collapsedLinkText)
            collapsedText = collapsedSpannable

            val expandedSpannable = SpannableStringBuilder()
                    .append(rawText)
                    .append(" ")
                    .append(expandedLinkText)
            if(toggleByLinkOnly) {
                expandedSpannable.setSpan(clickableSpan, expandedSpannable.length - (expandedLinkText.length), expandedSpannable.length, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
            }
            onCustomizeSpannable?.invoke(expandedSpannable, expandedLinkText)
            expandedText = expandedSpannable
        }

        super.setText(if(isExpanded) expandedText else collapsedText, bufferType)

        movementMethod = LinkMovementMethod.getInstance()
        highlightColor = Color.TRANSPARENT

        setOnClickListener(if(toggleByLinkOnly) null else this)
    }

    override fun setText(text: CharSequence?, type: BufferType?) {
        rawText = text?: ""
        bufferType = type?: BufferType.NORMAL
        update()
    }

    override fun onClick(v: View?) {
        toggle()
    }

    fun toggle() {
        isExpanded = !isExpanded
        super.setText(if(isExpanded) expandedText else collapsedText, bufferType)
        onStateChange?.invoke(isExpanded)
    }

    private val clickableSpan = object: ClickableSpan() {

        override fun onClick(widget: View) {
            toggle()
        }

        override fun updateDrawState(ds: TextPaint) {
            ds.color = linkColor
        }
    }
}